src/beaconEmitter.ts

```diff
@@ -0,0 +1,191 @@
+import noble from '@abandonware/noble';
+import EventEmitter from 'events';
+import _ from 'lodash';
+
+import logger from './logger';
+import { Beacon, Config, Peripherial } from './model';
+
+/**
+ * Base emitted class to emit high level detections for consumer apps.
+ */
+export default class BeaconEmitter extends EventEmitter {
+    /**
+     * A map of all recently detected beacons mapped by join code.
+     */
+    beacons: Map<string, Beacon>;
+
+    /**
+     * The config object used by the emitter.
+     */
+    config: Config;
+
+    /**
+     * List of previously reported beacons to match and avoid unnecesary reporting.
+     */
+    lastReportedBeacons: Beacon[];
+
+    /**
+     * The beacon last reported to be the best.
+     */
+    lastReportedBestBeacon: Beacon | undefined;
+
+    /**
+     * Timeout reference to start/stop reporting timer.
+     */
+    reporterTimer: any;
+
+    /**
+     * Instantiates a new {@code BeaconEmitter}.
+     *
+     * @param config - The config object to use
+     */
+    constructor(config?: {
+        beaconUUID?: string;
+        beaconDismissTimeoutSeconds?: number;
+        reportIntervalMillisecs?: number;
+    }) {
+        super();
+
+        this.beacons = new Map();
+        this.config = new Config(config);
+        this.lastReportedBeacons = [];
+
+        // Binding this
+        this.onDiscover = this.onDiscover.bind(this);
+        this.onScanStart = this.onScanStart.bind(this);
+        this.onScanStop = this.onScanStop.bind(this);
+        this.reportBeacons = this.reportBeacons.bind(this);
+
+        // Setting event listeners
+        noble.on('discover', this.onDiscover);
```

Comment:
is it ok to se these here so early?

Reply:
I don't see any issues with it. Generally my approach to event listeners is to set them as early as possible. Why are you concerned?

Reply:
Not concerned, just wondered if there was a problem doing it early. If there isn't then no problemo!